pkg/backends/calico/calico_test.go

```diff
@@ -0,0 +1,146 @@
+package calico
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/cache"
+)
+
+func addEndpointSubset(ep *v1.Endpoints, nodename string) {
+	ep.Subsets = append(ep.Subsets, v1.EndpointSubset{
+		Addresses: []v1.EndpointAddress{
+			v1.EndpointAddress{
+				NodeName: &nodename}}})
+}
+
+var _ = Describe("RouteGenerator", func() {
+	var (
+		rg   *routeGenerator
+		meta metav1.ObjectMeta
+		svc  *v1.Service
+		ep   *v1.Endpoints
+	)
+
+	BeforeEach(func() {
+		rg = &routeGenerator{
+			nodeName:    "foobar",
+			svcIndexer:  cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			epIndexer:   cache.NewIndexer(cache.MetaNamespaceKeyFunc, nil),
+			svcRouteMap: make(map[string]string),
+			client: &client{
+				cache: make(map[string]string),
+			},
+		}
+		meta = metav1.ObjectMeta{Namespace: "foo", Name: "bar"}
+		svc = &v1.Service{
+			ObjectMeta: meta,
+			Spec: v1.ServiceSpec{
+				Type:                  v1.ServiceTypeClusterIP,
+				ClusterIP:             "127.0.0.1",
+				ExternalTrafficPolicy: v1.ServiceExternalTrafficPolicyTypeLocal,
+			}}
+		ep = &v1.Endpoints{
+			ObjectMeta: meta,
+		}
+	})
+
+	Describe("Get one for the other", func() {
+		Context("getServiceForEndpoints", func() {
+			It("should work", func() {
+				// getServiceForEndpoints
+				rg.svcIndexer.Add(svc)
+				fetchedSvc, key := rg.getServiceForEndpoints(ep)
+				Expect(fetchedSvc.ObjectMeta).To(Equal(svc.ObjectMeta))
+				Expect(key).To(Equal("foo/bar"))
+			})
+		})
+		Context("getEndpointsForService", func() {
+			It("should work", func() {
+				// getEndpointsForService
+				rg.epIndexer.Add(ep)
+				fetchedEp, key := rg.getEndpointsForService(svc)
+				Expect(fetchedEp.ObjectMeta).To(Equal(ep.ObjectMeta))
+				Expect(key).To(Equal("foo/bar"))
+			})
+		})
+	})
+
+	Describe("(un)setRouteForSvc", func() {
+		BeforeEach(func() {
+			addEndpointSubset(ep, rg.nodeName)
+		})
+		Context("svc = svc, ep = nil", func() {
+			It("should work", func() {
```

Comment:
Let's put some more detailed It() statements in here.

e.g. `It("should set an unset routes for a service")`